### DISCUSSION TITLE
How do I start mDNS service over an Ethernet interface on Arduino-ESP32

### DISCUSSION THREAD

**Discussion by avillacis** [espressif/arduino-esp32#7213] [Q&A]
I am currently experimenting on connecting an Ethernet chipset (W5500) to an ESP32-S2 board using SPI. So far, DHCP runs over the interface, connection is established, and I can view a HTTP server formerly visible through the WiFi STA and AP interfaces. However, my starting project was supposed to enable mDNS using the ESPmDNS class, which I see is a wrapper over the mdns APIs from Espressif. My project correctly announced _http._tcp availability over the STA interface, but the problem is that it is not doing so over the Ethernet interface. Why? Do I have to do some extra step to enable mDNS over Ethernet?

I have tried disassociating the STA interface from the local access point, then rebooting. In this scenario, the project sets up the AP interface only, and now the Ethernet interface. And it still does not announce the mDNS service.

**Comment by MathewHDYT** [comment]
The ethernet library is explained a little bit more in this [pull request](https://github.com/espressif/arduino-esp32/pull/7163).

It contains this picture, the main point is that the `Ethernet` used by the `Arduino` is not the same `Ethernet` used by the `Espressif IDF`, base class. Meaning that you probably are not establishing a connection with the underlying `Espressif IDF` Ethernet Driver but the external Arduino Library.

This is the case because the `W5500` is currently only supported when using the external connection, which uses a completely different interface. Mainly because the `Arduino` library existed before the base `Espressif IDF` had support for the `W5500`.

![image](https://user-images.githubusercontent.com/48954742/213907328-b0aae48b-e24b-4f84-a8c1-9e6032e6ca67.png)

-----------------

There are few recommendations I could make, the easiest would probably be waiting until the pull request might be merged or if you do not require the `SPI`, you can copy [this method](https://github.com/espressif/arduino-esp32/pull/7163/files#diff-ac446ab01e7ad1e517b6da5f710759cf9f867d668175edc0f988aa032a0f8c94R412) into your code and use it to initialise a Ethernet connection with the base `Espressif IDF`.

Be aware this will cause unexpected errors if you use the `SPI? with `Arduio` (SD Card, External Display, Ethernet), because the `Espressif IDF` and `Arduino` use their own `SPI` driver. Which results in the SPI not working if both the `Espressif IDF` and `Arduino` each attempt to send data over the `SPI` driver.

**Comment by avillacis** [comment]
I should have been more explicit in my scenario.

I am not using the Ethernet class that comes with Arduino-ESP32, precisely because it does not hook W5500 support.

Also, I am not currently using any external Arduino Library to manage the W5500. I used one in the past to check that the W5500 wiring was actually working, but then I implemented my own initialization for the W5500. My initialization is supposed to do pretty much what the merge request is doing, based around the calls to `esp_eth_mac_new_w5500()` and `esp_eth_phy_new_w5500()`. Both of these methods are exposed in the ESP32 libraries shipped in the Arduino-ESP32 2.0.6 library. I end the initialization with calls to `esp_netif_attach()` and `esp_eth_start()`, just as in the merge request.

However, despite doing all that, and (presumably) adding the interface to the list of interfaces known by ESP-IDF, MDNS does not work through the interface, contrary to what I expected.

**Comment by MathewHDYT** [comment]
I seem to have found an [issue](https://github.com/espressif/esp-idf/issues/10359) that tries to do the same thing just with barebones `ESP-IDF` components. It seems to additionaly link to this [example](https://github.com/espressif/esp-protocols/blob/master/components/mdns/examples/main/mdns_example_main.c), which might be useful as well.
